src/python/py/models/builder.py

```diff
@@ -1495,6 +1506,125 @@ def make_mlp_fc(self, layer_id, mlp, root_input):
+        #                 /       \
+        #         router_MatMul    |
+        #             /     \      |
+        #         Shape      |     |
+        #           |        |     |
+        #         Gather     |     |
+        #           |        |     |
+        #       Unsqueeze    |     |
+        #           |        |    /
+        #        Concat     /    /
+        #             \    /    /
+        #             Reshape  /
+        #                 \   /
+        #                  MoE
+        #                   |
+        #                 output
+        num_experts = self.moe_attrs["num_experts"]
+        top_k = self.moe_attrs["top_k"]
+        activation_type = self.moe_attrs["activation_type"]
+        normalize_routing_weights = self.moe_attrs["normalize_routing_weights"]
+        use_sparse_mixer = self.moe_attrs["use_sparse_mixer"]
+        use_int4 = self.moe_attrs["use_int4"]
+
+        gate_ops_base = f"/model/layers.{layer_id}/moe/gate/"
+        moe_name = f"/model/layers.{layer_id}/moe"
+
+        # Make MoE nodes
+        gate_name = gate_ops_base + "MatMul"
+        self.make_matmul(bsm.gate, gate_name, root_input)
+
+        shape_name = gate_ops_base + "Shape"
+        self.make_shape(shape_name, f"{gate_name}/output_0", shape=[3])
+
+        gather_name = gate_ops_base + "Gather"
+        self.make_gather(gather_name, [f"{shape_name}/output_0", "/model/constants/TensorProto.INT64/0D/2"], axis=0)
+
+        unsqueeze_name = gate_ops_base + "Unsqueeze"
+        self.make_unsqueeze(unsqueeze_name, [f"{gather_name}/output_0", "/model/constants/TensorProto.INT64/1D/0"], dtype=TensorProto.INT64, shape=[1])
+
+        concat_name = gate_ops_base + "Concat"
+        self.make_concat(concat_name, ["/model/constants/TensorProto.INT64/1D/-1", f"{unsqueeze_name}/output_0"], dtype=TensorProto.INT64, shape=[2], axis=0)
+
+        gate_reshape_name = gate_ops_base + "Reshape"
+        self.make_reshape(gate_reshape_name, [f"{gate_name}/output_0", f"{concat_name}/output_0"], dtype=self.io_dtype, shape=['num_rows', num_experts])
+
+        def quant_dequant(weights, quant_mode: bool = True):
+            type = torch.quint4x2 if quant_mode else torch.int8
+            # The following line is needed use torch.ops.trtllm._symmetric_quantize_last_axis_of_batched_matrix()
+            import tensorrt_llm
+
+            _, processed_q_weight, torch_weight_scales = (
+                torch.ops.trtllm._symmetric_quantize_last_axis_of_batched_matrix(weights.T.cpu().contiguous(), type)
+            )
+
+            return torch_weight_scales.to(torch.float16), processed_q_weight
+
+        w1_list = []
+        w2_list = []
+        w3_list = []
+        w1_scale_list = []
+        w2_scale_list = []
+        w3_scale_list = []
+
+        for i in range(num_experts):
+            # Quantize the weights with uint8
+            w1_scale, pre_qweight1= quant_dequant(bsm.experts[i].w1.weight, use_int4)
```

Comment:
Is `QMoE` the only op available to create? Can a regular `MoE` op also be created if the user wants the non-quantized version?

Reply:
Will add MoE when GenAI tool supports multi-gpu